### DISCUSSION TITLE
classes/class_json

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#248] [User-contributed notes]
# classes/class_json

Inherits: Resource< RefCounted< Object Helper class for creating and parsing JSON data. Description: The JSON class enables all data types to be converted to and from a JSON string. This is useful ...

https://docs.godotengine.org/en/stable/classes/class_json.html

**Comment by 8-bit-pixel** [comment]
Will it ever have built-in support for YAML or JSON5?
Regular JSON and XML aren’t so human friendly, and having to write out large chunks of NPC dialogue in one of those would be quite labour-intensive. Also, the comments provided by JSON5 and YAML are really useful for readability.

**Reply by tetrapod00** [reply]
YAML was proposed and rejected in https://github.com/godotengine/godot-proposals/issues/941 (see also https://github.com/godotengine/godot-proposals/issues/5108).
As for JSON5, I don't even see a proposal for it: https://github.com/godotengine/godot-proposals/issues?q=is%3Aissue+json5+.

**Reply by Calinou** [reply]
You can manually write files in the [ConfigFile](https://docs.godotengine.org/en/stable/classes/class_configfile.html) format, which provides usability similar to JSON5 and can store comments.

**Comment by schaefchenaufderweide** [comment]
I am sorry, but this help is such a complete mess. 

```gdscript
var data_to_send = ["a", "b", "c"]
var json_string = JSON.stringify(data_to_send)
# Save data
# ...
# Retrieve data
var json = JSON.new()
var error = json.parse(json_string)
if error == OK:
	var data_received = json.data
	if typeof(data_received) == TYPE_ARRAY:
		print(data_received) # Prints array
	else:
		print("Unexpected data")
else:
	print("JSON Parse Error: ", json.get_error_message(), " in ", json_string, " at line ", json.get_error_line())
```

Why would you first save data and then load it again? This doesnt make sense. Examples should state both loading and saving.

**Reply by ducviet321** [reply]
It's an example to save to a persistent storage, and then read it again when you reopen your app. Maybe the comment should be changed to `# Save data to persistent storage`

**Comment by schaefchenaufderweide** [comment]
Also this line doesnt make any sense:

```gdscript
if error == OK:
	var data_received = json.data
	if typeof(data_received) == TYPE_ARRAY:
		print(data_received) # Prints array
	else:
		print("Unexpected data")
```

Why would you ask if the data is an array, if you want a json-file?

**Reply by ducviet321** [reply]
Because in the example, it is an array. It was deserialized back into program object of type array

**Comment by PrimisSolutions** [comment]
A common use case here is to read JSON from a file, which the documentation here does not address.

Suppose you have a file containing a list of names, split into first and last names at res://Data/names.json:

```json
{
    "file_name": "names.json"
    "first_names" : ["Arthur", "James",  "Annie", "Jane"]
    "last_names" : ["Smith", "Edwards", "Thomas"]
}
```

```gdscript
# get_names.gd
    # Get name data in the form of a dictionary
	var json = JSON.new()
	var names_string= FileAccess.get_file_as_string("res://Data/names.json")
	var names_as_dict= JSON.parse_string(names_string)

    var example_name = names_as_dict["first_names"].pick_random() + " " + names_as_dict["last_names"].pick_random()
```

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot/pull/104132

**Reply by OtherRealms** [reply]
thank you, I could not have figure that out

**Comment by RedpingDev** [comment]
```gdscript
var data_to_send = ["a", "b", "c"]
var json_string = JSON.stringify(data_to_send)
# Save data
# ...
# Retrieve data
var json = JSON.new()
var error = json.parse(json_string)
if error == OK:
	var data_received = json.data
	if typeof(data_received) == TYPE_ARRAY:
		print(data_received) # Prints array
	else:
		print("Unexpected data")
else:
	print("JSON Parse Error: ", json.get_error_message(), " in ", json_string, " at line ", json.get_error_line())
```
Now, shouldn't we be doing JSON.new() here?
like this?
```gdscript
var data_to_send = ["a", "b", "c"]
var json_string = JSON.stringify(data_to_send)
# Save data
# ...
# Retrieve data
var parse_result = JSON.parse_string(json_string)
if parse_result.error == OK:
    var data_received = parse_result.result
    if typeof(data_received) == TYPE_ARRAY:
        print(data_received) # Prints array
    else:
        print("Unexpected data")
else:
    print("JSON Parse Error: ", parse_result.error_string)
```

**Comment by hsw2201** [comment]
So I don't see it anywhere in the docs but it seems
var my_data = preload("res://my_json.json").data
is easy way to get data from json file in disk I guess

**Reply by Calinou** [reply]
Note that this only works for JSON files that are imported as Resources in the project. This won't work for arbitrary JSON files on the filesystem (e.g. in `user://`).

**Comment by CreativeHouseDotOrg** [comment]
> [!WARNING]
> The docs say stringify() will convert **any data type** which is <ins>not true</ins>. Stringify only accepts Variants to be converted to a json formatted string.

> stringify() is used to convert <ins>**any data type**</ins> into a JSON string.

```cs

public static string Stringify(Variant data, string indent = "", bool sortKeys = true, bool fullPrecision = false)
{
  return NativeCalls.godot_icall_4_708(MethodBind0, data, indent, sortKeys.ToGodotBool(), fullPrecision.ToGodotBool());
}

```

**Reply by Calinou** [reply]
A Variant is technically "any data type", since even `null` is considered a Variant (one that is of `Nil` type).

**Reply by CreativeHouseDotOrg** [reply]
Yeah but good luck trying that with class objects other than Variants like dtos.

To be able to serialize custom objects I started using `System.Text.Json.JsonSerializer.Serialize` and `System.Text.Json.JsonSerializer.Deserialize<T>`.

And imo "any data type" is to ambigious since variants are variants and "any" should really be referred to "any" including variants AND custom objects.

**Reply by EiTaNBaRiBoA** [reply]
> Yeah but good luck trying that with class objects other than Variants like dtos.
> 
> To be able to serialize custom objects I started using `System.Text.Json.JsonSerializer.Serialize` and `System.Text.Json.JsonSerializer.Deserialize<T>`.
> 
> And imo "any data type" is to ambigious since variants are variants and "any" should really be referred to "any" including variants AND custom objects.

I’ve developed JsonClassConverter, a Godot plugin designed serializing and deserializing Godot classes to and from JSON, and it has features for JSON comparison, diffing, and manipulation. 
[ForgeJSONGD Repo](https://github.com/EiTaNBaRiBoA/ForgeJSONGD)
[ForgeJSONGD in Godot Asset Library](https://godotengine.org/asset-library/asset/3073)

**Reply by CreativeHouseDotOrg** [reply]
> > Yeah but good luck trying that with class objects other than Variants like dtos.
> > To be able to serialize custom objects I started using `System.Text.Json.JsonSerializer.Serialize` and `System.Text.Json.JsonSerializer.Deserialize<T>`.
> > And imo "any data type" is to ambigious since variants are variants and "any" should really be referred to "any" including variants AND custom objects.
> 
> I’ve developed JsonClassConverter, a Godot plugin designed serializing and deserializing Godot classes to and from JSON, and it has features for JSON comparison, diffing, and manipulation. [JsonClassConverter Repo](https://github.com/EiTaNBaRiBoA/JsonClassConverter) [JsonClassConverter in Godot Asset Library](https://godotengine.org/asset-library/asset/3073)

Is there a C# Wrapper as well?

**Reply by EiTaNBaRiBoA** [reply]
> > > Yeah but good luck trying that with class objects other than Variants like dtos.
> > > To be able to serialize custom objects I started using `System.Text.Json.JsonSerializer.Serialize` and `System.Text.Json.JsonSerializer.Deserialize<T>`.
> > > And imo "any data type" is to ambigious since variants are variants and "any" should really be referred to "any" including variants AND custom objects.
> > 
> > 
> > I’ve developed JsonClassConverter, a Godot plugin designed serializing and deserializing Godot classes to and from JSON, and it has features for JSON comparison, diffing, and manipulation. [JsonClassConverter Repo](https://github.com/EiTaNBaRiBoA/JsonClassConverter) [JsonClassConverter in Godot Asset Library](https://godotengine.org/asset-library/asset/3073)
> 
> Is there a C# Wrapper as well?

Yes
```CSharp
using Godot;

[GlobalClass]
public partial class PlayerData : GodotObject
{
    [Export]
    public string Name { get; set; }

    [Export]
    public int Score { get; set; }
}

public partial class GameManager : Node
{
    private const string SavePath = "user://player_data.json";
    private const string ForgeJSONGDPath = "res://addons/ForgeJSONGD/ForgeJSONGD.gd";

    public override void _Ready()
    {
        // Load the ForgeJSONGD singleton script
        var forgeJsonScript = ResourceLoader.Load<Script>(ForgeJSONGDPath);

        // --- SAVE DATA ---
        var player = new PlayerData { Name = "C-Sharp Hero", Score = 1500 };
        var jsonDict = forgeJsonScript.Call("class_to_json", player);
        forgeJsonScript.Call("store_json_file", SavePath, jsonDict);
        GD.Print("Player data saved from C#!");

        // --- LOAD DATA ---
        var loadedPlayer = forgeJsonScript.Call("json_file_to_class", new PlayerData(), SavePath).As<PlayerData>();
        if (loadedPlayer != null)
        {
            GD.Print($"Loaded Player: {loadedPlayer.Name}, Score: {loadedPlayer.Score}");
            // Output: Loaded Player: C-Sharp Hero, Score: 1500
        }
    }
}
```

**Comment by Nivdc** [comment]
The description of the `stringify` function says:

> Note: The JSON specification doesn't define integer or float types, but only a number type. Therefore, converting a Variant to JSON text will convert all numerical values ​​to float types. 
﻿

But I found that this is incorrect in version 4.5.1:
```gdscript
print(JSON.stringify(65)) # output: 65
print(JSON.parse_string("65")) # output: 65.0
print(JSON.stringify(JSON.parse_string(JSON.stringify(65)))) # output: 65.0
```
﻿
It seems the `stringify`  doesn't change the type of the value, while `parse_string` does.
If this is intentional, please consider updating the document.
